packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp

```diff
@@ -277,8 +278,8 @@ class Reader : public Teuchos::Reader {
         break;
       }
       case Teuchos::YAML::PROD_BMAP_FSEQ: {
-        TEUCHOS_ASSERT(rhs.at(4).type() == typeid(Array<Scalar>) ||
-            rhs.at(4).type() == typeid(Array<Array<Scalar> >));
+        TEUCHOS_ASSERT(rhs.at(4).type() == typeid(std::vector<Scalar>) ||
```

Comment:
@ibaned, why are all of these being switched over from Teuchos::Array to std::vector?  Teuchos::Array just uses std::vector in non-debug checking mode but has very strong runtime checking in a debug-mode build.

Reply:
I just saw the comment "I also did some unrelated cleanup" in the description field.

Unless there is some technical reason to use std::vector instead of Teuchos::Array, the usage of Teuchos::Array should be kept for safety.  Even in this context, std::vector::operator[]() is completely unchecked.  And we can't turn on checked STL because we need to link against C++ TPLs that are built without checked STL turned on.  For the motivation for using Teuchos::Array over raw std::vector, see section "5.5.3 Teuchos::Array<T>" in:

* https://trilinos.org/docs/dev/packages/teuchos/doc/html/TeuchosMemoryManagementSAND.pdf

If you have questions about this, then let's talk.

Reply:
The motivation was to avoid having to define `std::ostream& operator<<(std::ostream&, Scalar const&)` and `std::ostream& operator<<(std::ostream&, std::vector<Scalar>&)`. It is possible to access `std::vector` in a checked way with `std::vector::at()`, I can make sure to use that if thats the main issue. Otherwise if `Teuchos::Array` still sounds better then I can put all that back.

Reply:
The problem with `std::vector::at()` is that it always checks.   There is no portable way to turn on and off array bounds checking with std::vector due to C++ TPLs that need to be linked in.

What is the issue with `operator<<()` exactly?

Reply:
It just seems like unnecessary boilerplate (recall the same issue with `Teuchos::any`). I used to have empty implementations for both. I can put them back just to get `Teuchos::Array` working just to have toggle-able bounds checking. Most of `TeuchosParser` already unconditionally calls `std::vector::at` since I don't consider it performance-critical.

Reply:
`std::vector::at()` does not completely make `std::vector` safe. YOu can also easily over-step and increment `std::vector` iterator returned from `std::vector::begin()`.  The motivation for `Teuchos::Array` as an off-by-one error I made in a code several years ago that segfaulted in a very strange place.  When I ran valgrind on it, it segfaulted inside of valgrind.  It took me several days to find the off-by-one error with std::vector and I only found it by staring at the code for days.  Teuchos::Array would have caught that defect in 1 sec.